contracts/src/provers/HashiProver.sol

```diff
@@ -0,0 +1,111 @@
+        uint256 dstChainId;
+        /// @dev The RLP-encoded block from which we want to retrieve its hash from Hashi
+        bytes rlpEncodedBlockHeader;
+        /// @dev Parameters needed to validate the authenticity of a specified storage location in `RIP7755Inbox` on
+        /// the destination chain
+        StateValidator.AccountProofParameters dstAccountProofParams;
+    }
+
+    /// @notice This error is thrown when fulfillmentInfo.timestamp is less than request.finalityDelaySeconds from
+    /// current destination chain block timestamp.
+    error FinalityDelaySecondsInProgress();
+
+    /// @notice This error is thrown when verification of the authenticity of the `RIP7755Inbox` storage on the
+    /// destination chain fails
+    error InvalidStorage();
+
+    /// @notice This error is thrown when verification of proof.blockHash agaist the one stored in Hashi fails
+    error InvalidBlockHeader();
+
+    /// @notice This error is thrown when the number of bytes to convert into an uin256 is greather than 32
+    error BytesLengthExceeds32();
+
+    /// @notice Validates storage proofs and verifies fulfillment
+    ///
+    /// @custom:reverts If storage proof invalid.
+    /// @custom:reverts If fulfillmentInfo not found at inboxContractStorageKey on request.inboxContract
+    /// @custom:reverts If fulfillmentInfo.timestamp is less than request.finalityDelaySeconds from current destination
+    /// chain block timestamp.
+    ///
+    ///
+    /// @param inboxContractStorageKey The storage location of the data to verify on the destination chain
+    /// `RIP7755Inbox` contract
+    /// @param fulfillmentInfo The fulfillment info that should be located at `inboxContractStorageKey` in storage
+    /// on the destination chain `RIP7755Inbox` contract
+    /// @param request The original cross chain request submitted to this contract
+    /// @param proof The proof to validate
+    function validateProof(
+        bytes memory inboxContractStorageKey,
+        RIP7755Inbox.FulfillmentInfo calldata fulfillmentInfo,
+        CrossChainRequest calldata request,
+        bytes calldata proof
+    ) external view {
+        if (block.timestamp - fulfillmentInfo.timestamp < request.finalityDelaySeconds) {
+            revert FinalityDelaySecondsInProgress();
+        }
+
+        RIP7755Proof memory proofData = abi.decode(proof, (RIP7755Proof));
+        // Set the expected storage key and value for the `RIP7755Inbox`
+        proofData.dstAccountProofParams.storageKey = inboxContractStorageKey;
+        proofData.dstAccountProofParams.storageValue = _encodeFulfillmentInfo(fulfillmentInfo);
+
+        (uint256 blockNumber, bytes32 stateRoot) = _extractBlockNumberAndStateRoot(proofData.rlpEncodedBlockHeader);
+
+        /// @notice The ShoyuBashi check should be performed within the PrecheckContract to ensure the correct ShoyuBashi is being used.
+        (address shoyuBashi) = abi.decode(request.precheckData, (address));
```

Comment:
Since this is taking over the precheckData field, does that mean we can't use a precheck condition for requests using a Hashi prover?

Reply:
As discussed in private, we will replace `prececheckData` with a new field called `extraData`, which will be added.